### PR TITLE
Added fallbacks logic

### PR DESCRIPTION
**Pull request by suriksarkisyan**

### CODE REVIEW

File: sdk/src/main/java/com/qonversion/android/sdk/dto/QFallbackObject.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import com.qonversion.android.sdk.dto.products.QProduct
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+
+@JsonClass(generateAdapter = true)
+internal data class QFallbackObject(
+    @Json(name = "products") val products: Map<String, QProduct> = mapOf(),
+    @Json(name = "offerings") val offerings: QOfferings?,
+    @Json(name = "products_permissions") val productPermissions: Map<String, List<String>>? = null,
+    @Json(name = "remote_config_list") val remoteConfigList: QRemoteConfigList?,
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/dto/QFallbackObject.kt:13]: Do we need default values?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/QProductCenterManager.kt [hunk 2]
```diff
@@ -430,52 +430,63 @@ internal class QProductCenterManager internal constructor(
...
         }
 
-        launchResultCache.productPermissions?.let {
-            val permissions = grantPermissionsAfterFailedRestore(
-                purchaseHistoryRecords,
-                launchResult.products.values,
-                it
+        val products = launchResultCache.getActualProducts() ?: run {
+            failLocallyGrantingRestorePermissionsWithError(
+                launchError ?: QonversionError(QonversionErrorCode.LaunchError)
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/QProductCenterManager.kt:442]: Let's rewrite it in a simple way - first get both fields, then check both on null and then once fail if any of them is null

File: sdk/src/main/java/com/qonversion/android/sdk/internal/QProductCenterManager.kt [hunk 3]
```diff
@@ -430,52 +430,63 @@ internal class QProductCenterManager internal constructor(
...
-            }
-
-            val permissions = grantPermissionsAfterFailedPurchaseTracking(
-                purchase,
-                purchasedProduct,
-                it
+        val productPermissions = launchResultCache.getProductPermissions() ?: run {
+            failLocallyGrantingPurchasePermissionsWithError(
+                purchaseCallback,
+                launchError ?: QonversionError(QonversionErrorCode.LaunchError)
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/QProductCenterManager.kt:472]: Same here

File: sdk/src/main/java/com/qonversion/android/sdk/internal/QRemoteConfigManager.kt [hunk 4]
```diff
@@ -89,7 +96,20 @@ internal class QRemoteConfigManager @Inject constructor(
...
+                        val baseRemoteConfigList = fallbackData?.remoteConfigList ?: run {
+                            fireToCallbacks(contextKey) { onError(error) }
+                            return@onError
+                        }
+
+                        val remoteConfig = if (contextKey != null) {
+                            baseRemoteConfigList.remoteConfigForContextKey(contextKey)
+                        } else {
+                            baseRemoteConfigList.remoteConfigForEmptyContextKey
+                        }
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/QRemoteConfigManager.kt:108]: Small point - if we swap branches, we can get rid of `!` in condition (would be a but clearer)

File: sdk/src/main/java/com/qonversion/android/sdk/internal/api/ApiErrorMapper.kt [hunk 5]
```diff
@@ -106,6 +108,7 @@ internal class ApiErrorMapper @Inject constructor(private val helper: ApiHelper)
             20008, 20010, 20203, 20210 -> QonversionErrorCode.PurchaseInvalid
             20011, 20012, 20013 -> QonversionErrorCode.ProjectConfigError
             20201 -> QonversionErrorCode.InvalidStoreCredentials
+            in Constants.INTERNAL_SERVER_ERROR_MIN ..Constants.INTERNAL_SERVER_ERROR_MAX -> QonversionErrorCode.InternalServerError
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/api/ApiErrorMapper.kt:111]: Don't we mix qonversion error codes with http codes here?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.InputStream
+import java.io.InputStreamReader
+import javax.inject.Inject
+
+internal class QFallbacksService @Inject constructor(
+    moshi: Moshi,
+    private val logger: Logger
+) {
+    private val jsonAdapter: JsonAdapter<QFallbackObject> = moshi.adapter(QFallbackObject::class.java)
+    fun obtainFallbackData(): QFallbackObject? {
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt:19]:
```suggestion
    private val jsonAdapter: JsonAdapter<QFallbackObject> = moshi.adapter(QFallbackObject::class.java)

    fun obtainFallbackData(): QFallbackObject? {
```

File: sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        return try {
+            val json: String = getStringFromFile("qonversion_fallbscks.json")
+            val fallbackData: QFallbackObject? = jsonAdapter.fromJson(json)
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt:22]: is result really nullable?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    fun getStringFromFile(filePath: String): String {
+        val file = File(filePath)
+        val fileInputStream = FileInputStream(file)
+        val result = convertStreamToString(fileInputStream)
+        fileInputStream.close()
+
+        return result
+    }
+
+}
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt:54]:
```suggestion
    }
}
```

File: sdk/src/main/java/com/qonversion/android/sdk/internal/utils.kt [hunk 9]
```diff
@@ -1,5 +1,10 @@
 package com.qonversion.android.sdk.internal
 
+import com.qonversion.android.sdk.dto.QonversionError
+import com.qonversion.android.sdk.dto.QonversionErrorCode
+
 internal val Int.daysToSeconds get() = this * 24L * 60 * 60
 
 internal val Int.daysToMs get() = daysToSeconds * 1000
+
+internal val QonversionError.shouldFireFallback get(): Boolean = this.code == QonversionErrorCode.NetworkConnectionFailed || this.code.
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/utils.kt:10]: `|| this.code. `?

File: sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt (hunk 7 above)
Comment (suriksarkisyan) [reply sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt:22]: Yes, it is

File: sdk/src/main/java/com/qonversion/android/sdk/dto/QonversionError.kt [hunk 10]
```diff
@@ -2,7 +2,8 @@ package com.qonversion.android.sdk.dto
 
 data class QonversionError(
     val code: QonversionErrorCode,
-    val additionalMessage: String = ""
+    val additionalMessage: String = "",
+    internal val httpCode: Int? = null
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/dto/QonversionError.kt:6]: We agreed, that you'll also remove purchase callback with HTTP code and use this field there

File: sdk/src/main/java/com/qonversion/android/sdk/QonversionConfig.kt [hunk 11]
```diff
@@ -75,6 +75,14 @@ class QonversionConfig internal constructor(
             this.entitlementsCacheLifetime = lifetime
         }
 
+        /**
+         * Sets fallback file identifier.
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/QonversionConfig.kt:79]: Lets add a couple of words about why should the file be used

File: sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt [hunk 12]
```diff
@@ -17,10 +17,11 @@ internal class QFallbacksService(
     private val logger: Logger
 ) {
     private val jsonAdapter: JsonAdapter<QFallbackObject> = moshi.adapter(QFallbackObject::class.java)
+    private val filePath = "qonversion_fallbacks.json"
```
Comment (SpertsyanKM) [inline sdk/src/main/java/com/qonversion/android/sdk/internal/services/QFallbacksService.kt:20]: It's local variable now, to make it constant move above the class name and rewrite as follows:
```
private const val FILE_PATH = "qonversion_fallbacks.json"
```

And let's name it FALLBACK_FILE_NAME? as it's not path?

**Review by SpertsyanKM — APPROVED** [review]

**Review by SpertsyanKM — APPROVED** [review]

**Pull request merged**